common/persistence/cassandra/cassandraPersistence.go

```diff
@@ -798,153 +794,539 @@ func (d *cassandraPersistence) CreateWorkflowExecution(
-				return nil, &p.CurrentWorkflowConditionFailedError{Msg: msg}
-			} else if rowType == rowTypeExecution && runID == executionInfo.RunID {
-				msg := fmt.Sprintf("Workflow execution already running. WorkflowId: %v, RunId: %v, rangeID: %v",
-					executionInfo.WorkflowID, executionInfo.RunID, request.RangeID)
-				lastWriteVersion := common.EmptyVersion
-				if previous["workflow_last_write_version"] != nil {
-					lastWriteVersion = previous["workflow_last_write_version"].(int64)
+			case conditionFailureErr.CurrentWorkflowConditionFailInfo != nil:
+				return nil, &p.CurrentWorkflowConditionFailedError{
+					Msg: *conditionFailureErr.CurrentWorkflowConditionFailInfo,
 				}
+			case conditionFailureErr.WorkflowExecutionAlreadyExists != nil:
 				return nil, &p.WorkflowExecutionAlreadyStartedError{
-					Msg:              msg,
-					StartRequestID:   executionInfo.CreateRequestID,
-					RunID:            executionInfo.RunID,
-					State:            executionInfo.State,
-					CloseStatus:      executionInfo.CloseStatus,
-					LastWriteVersion: lastWriteVersion,
+					Msg:              conditionFailureErr.WorkflowExecutionAlreadyExists.OtherInfo,
+					StartRequestID:   conditionFailureErr.WorkflowExecutionAlreadyExists.CreateRequestID,
+					RunID:            conditionFailureErr.WorkflowExecutionAlreadyExists.RunID,
+					State:            conditionFailureErr.WorkflowExecutionAlreadyExists.State,
+					CloseStatus:      conditionFailureErr.WorkflowExecutionAlreadyExists.CloseStatus,
+					LastWriteVersion: conditionFailureErr.WorkflowExecutionAlreadyExists.LastWriteVersion,
 				}
+			default:
+				// If ever runs into this branch, there is bug in the code either in here, or in the implementation of nosql plugin
+				err := fmt.Errorf("unsupported conditionFailureReason error")
+				d.logger.Error("A code bug exists in persistence layer, please investigate ASAP", tag.Error(err))
+				return nil, err
+			}
+		}
+		return nil, convertCommonErrors(d.client, "CreateWorkflowExecution", err)
+	}
+
+	return &p.CreateWorkflowExecutionResponse{}, nil
+}
+
+func (d *cassandraPersistence) prepareTimerTasksForWorkflowTxn(
+	domainID, workflowID, runID string,
+	timerTasks []p.Task,
+) ([]*nosqlplugin.TimerTask, error) {
+	var tasks []*nosqlplugin.TimerTask
+
+	for _, task := range timerTasks {
+		var eventID int64
+		var attempt int64
+
+		timeoutType := 0
+
+		switch t := task.(type) {
+		case *p.DecisionTimeoutTask:
+			eventID = t.EventID
+			timeoutType = t.TimeoutType
+			attempt = t.ScheduleAttempt
+
+		case *p.ActivityTimeoutTask:
+			eventID = t.EventID
+			timeoutType = t.TimeoutType
+			attempt = t.Attempt
+
+		case *p.UserTimerTask:
+			eventID = t.EventID
+
+		case *p.ActivityRetryTimerTask:
+			eventID = t.EventID
+			attempt = int64(t.Attempt)
+
+		case *p.WorkflowBackoffTimerTask:
+			eventID = t.EventID
+			timeoutType = t.TimeoutType
+
+		case *p.WorkflowTimeoutTask:
+			// noop
```

Comment:
do we need to set the eventID and timeoutType?

Reply:
I believe it's not needed as the existing behavior: https://github.com/uber/cadence/blob/5848c942aa51ce34c52a01439188117b1e615a93/common/persistence/cassandra/cassandraPersistenceUtil.go#L929